File: Dynamic-Programming/64. Minimum Path Sum/minimumPathSum.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var minPathSum = function (grid) {
    let n = grid.length;
    let m = grid[0].length;

    // update first row
    for (let i = 1; i < m; i++) {
        grid[0][i] += grid[0][i - 1];
    }

    // update first column
    for (let i = 1; i < n; i++) {
        grid[i][0] += grid[i - 1][0];
    }

    // update point by point
    for (let i = 1; i < n; i++) {
        for (let j = 1; j < m; j++) {
            grid[i][j] += Math.min(grid[i - 1][j], grid[i][j - 1]);
        }
    }

    // console.log(grid, n, m)
    return grid[n - 1][m - 1];
};

File: Dynamic-Programming/64. Minimum Path Sum/minimumPathSum.java
// S(x, y)
// S(2, 2)
// option1 = S(1, 2) ?
// option2 = S(2, 1) ?
// return Math.min(option1, option2) + grid[2][2]
class Solution {
    public int minPathSum(int[][] grid) {
        int sizeX = grid[0].length;
        int sizeY = grid.length;
        
        SlidingMatrix memo = new SlidingMatrix();
        memo.values = new int[sizeX];
        
        for (int y = 0; y < sizeY; y++) {
            for (int x = 0; x < sizeX; x++) {
                int solution = -1;
                solution = better(solution, memo.get(x - 1, y));
                solution = better(solution, memo.get(x, y - 1));
                solution = add(solution, grid[y][x]);
                memo.feed(x, y, solution);
            }
        }
        return memo.get(sizeX - 1, sizeY - 1);
    }
    
    private int better(int solution1, int solution2) {
        if (solution1 == -1 && solution2 == -1) {
            return -1;
        } else if (solution1 == -1) {
            return solution2;
        } else if (solution2 == -1) {
            return solution1;
        } else {
            return Math.min(solution1, solution2);
        }
    }
    
    private int add(int solution, int value){
        return solution == -1 ? value : solution + value;
    }
    
    class SlidingMatrix {
        int[] values;

        int get(int x, int y) {
            if (y < 0 || x < 0) {
                return -1;
            }
            return values[x];
        }

        // setter
        void feed(int x, int y, int value) {
            values[x] = value;
        }
    }
}

